# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/platforms/agnostic/DAV1DDecoder.cpp
# Commit: e1ec2362ecb9
# Full Hash: e1ec2362ecb9193ca4c274a708145c6386c75270
# Author: Alex Chronopoulos <achronop@gmail.com>
# Date: 2019-04-02 15:46:39
# Description:
#   Bug 1540231 - Stop dispatching a task to the TaskQueue when we are already in it. r=jya
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D25576
# ==============================================================================

diff -r a0a1f6c8e4d1 -r e1ec2362ecb9 dom/media/platforms/agnostic/DAV1DDecoder.cpp
--- a/dom/media/platforms/agnostic/DAV1DDecoder.cpp	Tue Apr 02 06:22:35 2019 +0000
+++ b/dom/media/platforms/agnostic/DAV1DDecoder.cpp	Mon Apr 01 22:05:48 2019 +0000
@@ -62,19 +62,24 @@
 }
 
 void DAV1DDecoder::ReleaseDataBuffer(const uint8_t* buf) {
-  // The release callback may be called on a
-  // different thread defined by third party
-  // dav1d execution. Post a task into TaskQueue
-  // to ensure mDecodingBuffers is only ever
-  // accessed on the TaskQueue
+  // The release callback may be called on a different thread defined by the
+  // third party dav1d execution. In that case post a task into TaskQueue to
+  // ensure that mDecodingBuffers is only ever accessed on the TaskQueue.
   RefPtr<DAV1DDecoder> self = this;
-  nsresult rv = mTaskQueue->Dispatch(
-      NS_NewRunnableFunction("DAV1DDecoder::ReleaseDataBuffer", [self, buf]() {
-        DebugOnly<bool> found = self->mDecodingBuffers.Remove(buf);
-        MOZ_ASSERT(found);
-      }));
-  MOZ_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv));
-  Unused << rv;
+  auto releaseBuffer = [self, buf] {
+    MOZ_ASSERT(self->mTaskQueue->IsCurrentThreadIn());
+    DebugOnly<bool> found = self->mDecodingBuffers.Remove(buf);
+    MOZ_ASSERT(found);
+  };
+
+  if (mTaskQueue->IsCurrentThreadIn()) {
+    releaseBuffer();
+  } else {
+    nsresult rv = mTaskQueue->Dispatch(NS_NewRunnableFunction(
+        "DAV1DDecoder::ReleaseDataBuffer", std::move(releaseBuffer)));
+    MOZ_DIAGNOSTIC_ASSERT(NS_SUCCEEDED(rv));
+    Unused << rv;
+  }
 }
 
 RefPtr<MediaDataDecoder::DecodePromise> DAV1DDecoder::InvokeDecode(
